runtime/jaktlib/file_iterator.jakt

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2022, Rob Ryan <rob@robryan.me>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+import path { Path, get_path_separator }
+import os { platform_fs, platform_module, Target }
+
+import platform_fs() {
+    is_directory
+    DirectoryIterator
+}
+
+class RecursiveFileIterator implements(ThrowingIterable<String>) {
+    private extension: String
+    private directory_list: [String]
+    private current_directory: DirectoryIterator?
+    
+    public function make(directory: String, extension: String) throws -> RecursiveFileIterator {
+        return RecursiveFileIterator(extension, directory_list: [directory], current_directory: None)
+    }
+
+    public function next(mut this) throws -> String?  {
+        if not .current_directory.has_value() {
+
+            if .directory_list.size() > 0 {
+                let path = .directory_list.pop()!
+
+                .current_directory = is_directory(path)
+
+                if not .current_directory.has_value() {
+                    return path
+                }
+            } else {
+                return None
+            }
+        }
+
+        let next = .current_directory!.next()
+
+        if next.has_value() {
+            mut builder = StringBuilder::create()
+            builder.append_string(.current_directory!.get_path())
+            builder.append(get_path_separator())
+            builder.append_string(next!.0)
+
+            if next!.1 {
+                .directory_list.push(builder.to_string())
+
+                return .next()
+            }
+
+            let path = Path::from_string(builder.to_string())
+            if path.extension() == .extension {
+                return path.to_string()
+            } else {
+                return .next()
+            }
+        } else if .directory_list.size() != 0 {
+            .current_directory = is_directory(path: .directory_list.pop()!)
```

Comment:
`is_directory` gives you a `DirectoryIterator`...?

Reply:
Maybe it could be better named, the idea being if it is a directory it will return a `DirectoryIterator` if not it will return `None`. So that the one `opendir` call can be used both to test if it is a directory and iterate over it if it is. I will fix the windows version as it would return `(filename, is_directory)` currently instead on `None` if it is a file and not a directory.

Reply:
Could this just be `DirectoryIterator::from_path()`?

Reply:
Sure, changed that.